contracts/asset-proxy/contracts/src/ERC1155Proxy.sol

```diff
@@ -0,0 +1,271 @@
+                    mstore(32, 0x0000002000000000000000000000000000000000000000000000000000000000)
+                    mstore(64, 0x0000001553454e4445525f4e4f545f415554484f52495a454400000000000000)
+                    mstore(96, 0)
+                    revert(0, 100)
+                }
+
+                // Construct Table #3 in memory, starting at memory offset 0.
+                // The algorithm below maps asset data from Table #1 and Table #2 to Table #3, while
+                // scaling the `values` (Table #2) by `amount` (Table #1). Once Table #3 has
+                // been constructed in memory, the destination erc1155 contract is called using this 
+                // as its calldata. This process is divided into four steps, below.
+
+                ////////// STEP 1/4 //////////
+                // Map relevant fields from assetData (Table #2) into memory (Table #3)
+                // The Contents column of Table #2 is the same as Table #3,
+                // beginning from parameter 3 - `offset to ids (*)`
+                // The offsets in these rows are offset by 32 bytes in Table #3.
+                // Strategy:
+                // 1. Copy the assetData into memory at offset 32
+                // 2. Increment by 32 the offsets to `ids`, `values`, and `data`
+
+                // Load offset to `assetData`
+                let assetDataOffset := calldataload(4)
+
+                // Load length in bytes of `assetData`, computed by:
+                // 4 (function selector)
+                // + assetDataOffset
+                let assetDataLength := calldataload(add(4, assetDataOffset))
+
+                // This corresponds to the beginning of the Data Area for Table #3.
+                // Computed by:
+                // 4 (function selector)
+                // + assetDataOffset
+                // + 32 (length of assetData)
+                calldatacopy(
+                    32,                         // aligned such that "offset to ids" is at the correct location for Table #3
+                    add(36, assetDataOffset),   // beginning of asset data contents
+                    assetDataLength             // length of asset data
+                )
+
+                // Increment by 32 the offsets to `ids`, `values`, and `data`
+                mstore(68, add(mload(68), 32))
+                mstore(100, add(mload(100), 32))
+                mstore(132, add(mload(132), 32))
+
+                // Record the address of the destination erc1155 asset for later.
+                let assetAddress := and(
+                    mload(36),
+                    0x000000000000000000000000ffffffffffffffffffffffffffffffffffffffff
+                )
+
+                ////////// STEP 2/4 //////////
+                // load amount and ensure it's non-zero (we divide by amount below)
+                let amount := calldataload(100)
+                if iszero(amount) {
```

Comment:
This is how we handle this case in the MAP:

```
// Revert if `amount` != 0 and multiplication resulted in an overflow 
if iszero(or(
    iszero(amount),
    eq(div(totalAmount, amount), amountsElement)
)) {
    // Revert with `Error("UINT256_OVERFLOW")`
    mstore(0, 0x08c379a000000000000000000000000000000000000000000000000000000000)
    mstore(32, 0x0000002000000000000000000000000000000000000000000000000000000000)
    mstore(64, 0x0000001055494e543235365f4f564552464c4f57000000000000000000000000)
    mstore(96, 0)
    revert(0, 100)
}
```

So we don't actually revert if amount is 0, instead we don't check for an overflow. I'd prefer switching to this for the sake of consistency.

Reply:
Ah, okay. That makes sense. So we revert only when `amount` is non-zero and there is an overflow. 
```
amount=0 | no overflow | amount=0 OR no overflow | iszero(amount=0 OR no overflow)
   1           1               1                       0
   1           0               1                       0
   0           1               1                       0
   0           0               0                       1
```
And then when `amount=0` the `totalAmount` is also zero.